module-system/sov-modules-core/src/storage/scratchpad.rs

```diff
@@ -460,6 +479,69 @@ impl<'a, C: Context> StateReaderAndWriter for AccessoryWorkingSet<'a, C> {
     }
 }
 
+/// Module for archival state
+pub mod archival_state {
+    use super::*;
+
+    /// Enum for tagging the specific type of storage.
+    enum ArchivalStore<S: Storage> {
+        /// JMT store
+        Jmt(S),
+        /// Accessory store
+        Accessory(S),
+    }
+
+    /// Archival working set
+    pub struct ArchivalWorkingSet<C: Context> {
+        delta: ArchivalStore<<C as Spec>::Storage>,
+        archival_version: Version,
+        witness: <<C as Spec>::Storage as Storage>::Witness,
```

Comment:
Why do we need witness in `ArchivalWorkingSet`?

Reply:
I think we don't, but my guess is to keep it compatible with `Storage` methods, which require `Witness` to be passed explicitly via parameters.

I personally would prefer, that witness would be kept internally to the ProverStorage and ZkStorage, and created new one together with Storage instance itself. But this change is defenetly out of the scope of this PR.

Now we have case of tight coupling, so all users of "storage" trait should maintain witness for it.

@preston-evans98 , what do you think? I vaguely recall several months ago I've asked this question already.